components/cu_transform/src/tree.rs

```diff
@@ -0,0 +1,1572 @@
+use crate::error::{TransformError, TransformResult};
+use crate::transform::{StampedTransform, TransformStore};
+use crate::velocity::VelocityTransform;
+use crate::velocity_cache::VelocityTransformCache;
+use cu29::clock::CuTime;
+use cu_spatial_payloads::Transform3D;
+use dashmap::DashMap;
+use petgraph::algo::dijkstra;
+use petgraph::graph::{DiGraph, NodeIndex};
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::ops::Neg;
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::time::{Duration, Instant};
+
+/// Trait for types that can compute their inverse transformation
+pub trait HasInverse<T: Copy + Debug + 'static> {
+    fn inverse(&self) -> Self;
+}
+
+impl HasInverse<f32> for Transform3D<f32> {
+    fn inverse(&self) -> Self {
+        self.inverse()
+    }
+}
+
+impl HasInverse<f64> for Transform3D<f64> {
+    fn inverse(&self) -> Self {
+        self.inverse()
+    }
+}
+
+/// The cache entry for a transform query
+#[derive(Clone)]
+struct TransformCacheEntry<T: Copy + Debug + 'static> {
+    /// The cached transform result
+    transform: Transform3D<T>,
+    /// The time for which this transform was calculated
+    time: CuTime,
+    /// When this cache entry was last accessed
+    last_access: Instant,
```

Comment:
needs to be based on robotclock as we can restore tasks and go back in time and the determinism should be preserved